sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs

```diff
@@ -334,6 +343,10 @@ public void AddLink(string traceparent, string? tracestate, IDictionary<string,
                 _links.Add(linkedActivity);
             }
 
+#if !NET5_0
+            [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(Activity))]
+            [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(DiagnosticActivity))]
```

Comment:
Can this be a bit more precise? `All` is usually a bit too much.
Also in this place it's additionally problematic as it introduces warnings like:

```console
ILC : AOT analysis warning IL3050: Azure.Core.Pipeline.DiagnosticScope.ActivityAdapter.Start(): Using member 'System.Enum.GetValues(Type)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. It might not be possible to create an array of the enum type at runtime. Use the GetValues<TEnum> overload or the GetValuesAsUnderlyingType method instead.
```

(There's a bug in the compiler - it should really point the warning to this line of code - currently it doesn't give it any location in code unfortunately).

The reason for this warning is a bit complicated:
* The `All` annotation on `Activity` marks all nested types on Activity
* Once of those types is an `enum`
* The `All` annotation then applies to `System.Enum` (since it's a base type of the nested type)
* This marks all methods on `System.Enum`
* `System.Enum.GetValues` is not AOT compatible - as per the warning

If we could scope the dependency down to basically anything else but `All` it would solve this problem.

Reply:
Updated!